### ISSUE TITLE
CI: Potential deadlock from xDS

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#9613]
It seems the following test run got a potential deadlock:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/2635/

possible reasons:
```
2019-11-14T16:21:43.806898899Z level=debug msg="current resource version is 53, waiting for it to become > 53" subsys=xds xdsAckedVersion=53 xdsClientNode=10.10.1.181 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2019-11-14T16:21:44.179255276Z level=debug msg="xDS stream canceled" error="rpc error: code = Canceled desc = context canceled" subsys=xds xdsStreamID=1
2019-11-14T16:21:44.179279825Z level=debug msg="xDS stream context canceled" error="context canceled" subsys=xds xdsStreamID=2
2019-11-14T16:21:44.179283601Z level=debug msg="xDS stream canceled" error="rpc error: code = Canceled desc = context canceled" subsys=xds xdsStreamID=2
2019-11-14T16:21:44.17928656Z level=info msg="xDS stream closed" subsys=xds xdsStreamID=1
2019-11-14T16:21:44.179289398Z level=info msg="Envoy: Closing access log connection" subsys=envoy-manager
2019-11-14T16:21:44.316151762Z level=debug msg="xDS stream canceled" error="rpc error: code = Canceled desc = context canceled" subsys=xds xdsStreamID=5
2019-11-14T16:21:44.316171882Z level=debug msg="xDS stream context canceled" error="context canceled" subsys=xds xdsStreamID=6
2019-11-14T16:21:44.316175745Z level=debug msg="xDS stream canceled" error="rpc error: code = Canceled desc = context canceled" subsys=xds xdsStreamID=6
2019-11-14T16:21:44.316178705Z level=debug msg="xDS stream context canceled" error="context canceled" subsys=xds xdsStreamID=5
```

```
2019-11-14T16:26:54.586355298Z POTENTIAL DEADLOCK:
2019-11-14T16:26:54.586386802Z Previous place where the lock was grabbed
2019-11-14T16:26:54.586392915Z goroutine 34318 lock 0xc0010964d0
2019-11-14T16:26:54.586445613Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock ??? <<<<<
2019-11-14T16:26:54.586455286Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:261 endpoint.(*Endpoint).regenerate ???
2019-11-14T16:26:54.586486042Z /go/src/github.com/cilium/cilium/pkg/endpoint/events.go:57 endpoint.(*EndpointRegenerationEvent).Handle ???
2019-11-14T16:26:54.586494044Z /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:256 eventqueue.(*EventQueue).Run.func1 ???
2019-11-14T16:26:54.586498859Z /usr/local/go/src/sync/once.go:44 sync.(*Once).Do ???
2019-11-14T16:26:54.586503137Z 
2019-11-14T16:26:54.586506975Z Have been trying to lock it again for more than 5m10s
2019-11-14T16:26:54.58651082Z goroutine 42146 lock 0xc0010964d0
2019-11-14T16:26:54.586547653Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock ??? <<<<<
2019-11-14T16:26:54.586556356Z /go/src/github.com/cilium/cilium/daemon/endpoint.go:593 main.(*Daemon).deleteEndpointQuiet ???
2019-11-14T16:26:54.586572849Z /go/src/github.com/cilium/cilium/daemon/endpoint.go:551 main.(*Daemon).deleteEndpoint ???
2019-11-14T16:26:54.586577492Z /go/src/github.com/cilium/cilium/daemon/endpoint.go:678 main.(*Daemon).DeleteEndpoint ???
2019-11-14T16:26:54.586580993Z /go/src/github.com/cilium/cilium/daemon/endpoint.go:694 main.(*deleteEndpointID).Handle ???
2019-11-14T16:26:54.586584905Z /go/src/github.com/cilium/cilium/api/v1/server/restapi/endpoint/delete_endpoint_id.go:65 endpoint.(*DeleteEndpointID).ServeHTTP ???
2019-11-14T16:26:54.586588706Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 middleware.NewOperationExecutor.func1 ???
2019-11-14T16:26:54.586593075Z /usr/local/go/src/net/http/server.go:1995 http.HandlerFunc.ServeHTTP ???
2019-11-14T16:26:54.58662697Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/router.go:76 middleware.NewRouter.func1 ???
2019-11-14T16:26:54.586635138Z /usr/local/go/src/net/http/server.go:1995 http.HandlerFunc.ServeHTTP ???
2019-11-14T16:26:54.586639234Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 middleware.Redoc.func1 ???
2019-11-14T16:26:54.586643052Z /usr/local/go/src/net/http/server.go:1995 http.HandlerFunc.ServeHTTP ???
2019-11-14T16:26:54.586646718Z /go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/spec.go:46 middleware.Spec.func1 ???
2019-11-14T16:26:54.586650648Z /usr/local/go/src/net/http/server.go:1995 http.HandlerFunc.ServeHTTP ???
2019-11-14T16:26:54.586654438Z /go/src/github.com/cilium/cilium/pkg/metrics/middleware.go:69 metrics.(*APIEventTSHelper).ServeHTTP ???
2019-11-14T16:26:54.58665829Z /go/src/github.com/cilium/cilium/pkg/api/apipanic.go:53 api.(*APIPanicHandler).ServeHTTP ???
2019-11-14T16:26:54.586661959Z /usr/local/go/src/net/http/server.go:2774 http.serverHandler.ServeHTTP ???
2019-11-14T16:26:54.586665421Z /usr/local/go/src/net/http/server.go:1878 http.(*conn).serve ???
2019-11-14T16:26:54.586668934Z 
2019-11-14T16:26:54.600750122Z Here is what goroutine 34318 doing now
2019-11-14T16:26:54.600768955Z goroutine 34318 [select, 5 minutes]:
2019-11-14T16:26:54.600772452Z github.com/cilium/cilium/pkg/completion.(*WaitGroup).Wait(0xc001a01d10, 0x0, 0x0)
2019-11-14T16:26:54.600775158Z 	/go/src/github.com/cilium/cilium/pkg/completion/completion.go:104 +0x190
2019-11-14T16:26:54.600777913Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).WaitForProxyCompletions(0xc0010962c0, 0xc001a01d10, 0x541a700, 0x0)
2019-11-14T16:26:54.600780487Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:402 +0x1d3
2019-11-14T16:26:54.600782976Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc0010962c0, 0xc001bd0280, 0x0, 0xc0023d3d00, 0x0, 0x0)
2019-11-14T16:26:54.600785563Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:439 +0x640
2019-11-14T16:26:54.600788042Z github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc0010962c0, 0xc001bd0280, 0x0, 0x0)
2019-11-14T16:26:54.600798344Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:335 +0xb29
2019-11-14T16:26:54.600800989Z github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc0017b9a30, 0xc0017c04e0)
2019-11-14T16:26:54.600803513Z 	/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:57 +0x467
2019-11-14T16:26:54.600805909Z github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run.func1()
2019-11-14T16:26:54.600808343Z 	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:256 +0x150
2019-11-14T16:26:54.600810893Z sync.(*Once).Do(0xc0019b2fd8, 0xc001c88740)
2019-11-14T16:26:54.60081333Z 	/usr/local/go/src/sync/once.go:44 +0xb3
2019-11-14T16:26:54.600815835Z created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
2019-11-14T16:26:54.600818393Z 	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0xaa
2019-11-14T16:26:54.600820886Z Other goroutines holding locks:
2019-11-14T16:26:54.600823332Z goroutine 34318 lock 0xc001a01d28
2019-11-14T16:26:54.600864891Z /go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock ??? <<<<<
2019-11-14T16:26:54.600871184Z /go/src/github.com/cilium/cilium/pkg/completion/completion.go:98 completion.(*WaitGroup).Wait ???
2019-11-14T16:26:54.600873846Z /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:402 endpoint.(*Endpoint).WaitForProxyCompletions ???
2019-11-14T16:26:54.600876404Z /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:439 endpoint.(*Endpoint).regenerateBPF ???
2019-11-14T16:26:54.600878843Z /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:335 endpoint.(*Endpoint).regenerate ???
2019-11-14T16:26:54.600881336Z /go/src/github.com/cilium/cilium/pkg/endpoint/events.go:57 endpoint.(*EndpointRegenerationEvent).Handle ???
2019-11-14T16:26:54.600883784Z /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:256 eventqueue.(*EventQueue).Run.func1 ???
2019-11-14T16:26:54.600886263Z /usr/local/go/src/sync/once.go:44 sync.(*Once).Do ???
2019-11-14T16:26:54.60088869Z 
2019-11-14T16:26:54.600891123Z 
2019-11-14T16:26:54.60089342Z 
```

PR where the test was seen:
https://github.com/cilium/cilium/pull/9610

Full archive: (See logs from `test_results/111-/K8sServicesTest_Checks_service_across_nodes_with_L7_policy_Tests_NodePort_with_L7_Policy`)

[test_results_Cilium-PR-Ginkgo-Tests-K8s_2635_BDD-Test-k8s-1.11.zip](https://github.com/cilium/cilium/files/3847982/test_results_Cilium-PR-Ginkgo-Tests-K8s_2635_BDD-Test-k8s-1.11.zip)

**Comment by jrajahalme** [comment]
Just based on the above it looks like Envoy is not responding with neither ACK nor NACK, which stalls the regeneration of the endpoint, which then blocks the deletion of the endpoint. The reason for Envoy not responding seems to be that the listener has been removed, which happens when the last redirect is removed.

It is possible that this is simply a race condition between xDS N/ACK and listener removal that typically does not fire, but has here. Will verify if this is the case.

**Comment by aanm** [comment]
Potentially fixed already #9635
